version: '3.9'

services:
  django:
    build: .
    container_name: django
    command: gunicorn --bind 0.0.0.0:8002 config.wsgi:application
    entrypoint: ["/entrypoint.sh"]
    volumes:
      - .:/app
      - db-data:/app/db_data
    expose:
      - "8002"
    networks:
      - app-network
    environment:
      - SECRET_KEY=${SECRET_KEY}
      - REDIS_URL=${REDIS_URL}
    healthcheck:
      test: [ "CMD", "curl", "-f", "http://localhost:8002" ]
      interval: 30s
      timeout: 10s
      retries: 3

  django_websocket:
    build: .
    container_name: django_websocket
    command: daphne -b 0.0.0.0 -p 8003 config.asgi:application
    entrypoint: ["/entrypoint.sh"]
    volumes:
      - .:/app
      - db-data:/app/db_data
    expose:
      - "8003"
    networks:
      - app-network
    environment:
      - SECRET_KEY=${SECRET_KEY}
      - REDIS_URL=${REDIS_URL}
    healthcheck:
      test: [ "CMD", "curl", "-f", "http://localhost:8003" ]
      interval: 30s
      timeout: 10s
      retries: 3

  nginx:
    image: nginx:latest
    container_name: nginx
    ports:
      - "8080:8080"
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf
      - ./nginx_logs:/var/log/nginx
    depends_on:
      django:
        condition: service_healthy
      django_websocket:
        condition: service_healthy
    links:
      - django
      - django_websocket
    networks:
      - app-network
    tmpfs:
      - /tmp/nginx:exec,mode=1777

networks:
  app-network:
    driver: bridge

volumes:
  db-data:
    driver: local
    driver_opts:
      type: none
      o: bind
      device: ./db_data
